src/services/fetch/web_sys.rs

```diff
@@ -315,156 +316,200 @@ impl FetchService {
+        let response = self.get_response(promise).await?;
+        let data = self.get_data(&response).await?;
+        Ok((data, response))
+    }
+
+    // Prepare the response callback.
+    // Notice that the callback signature must match the call from the javascript
+    // side. There is no static check at this point.
+    fn callback(&self, data: Result<DATA, anyhow::Error>, status: u16, headers: Option<Headers>) {
+        let mut response_builder = Response::builder().status(status);
+        if let Some(headers) = headers {
+            for (key, value) in header_iter(headers) {
+                response_builder = response_builder.header(key.as_str(), value.as_str());
+            }
+        }
+
+        // Deserialize and wrap response data into a Text or Binary object.
+        let response = response_builder
+            .body(OUT::from(data))
+            .expect("failed to build response, please report");
+        *self.active.borrow_mut() = false;
+        self.callback.emit(response);
+    }
+
+    async fn get_response(&self, fetch_promise: Promise) -> Result<WebResponse, FetchError> {
+        let response = JsFuture::from(fetch_promise).await.map_err(|err| {
+            let dom_exception = DomException::from(err);
+            match dom_exception.code() {
+                DomException::ABORT_ERR => FetchError::Canceled,
+                DomException::NETWORK_ERR => FetchError::NetworkFailure,
+                _ => FetchError::FetchFailed,
+            }
+        })?;
+
+        if *self.active.borrow() {
+            Ok(WebResponse::from(response))
+        } else {
+            Err(FetchError::Canceled)
+        }
+    }
+
+    async fn get_data(&self, response: &WebResponse) -> Result<DATA, FetchError> {
+        let data_promise = if self.binary {
+            response.array_buffer()
+        } else {
+            response.text()
+        }
+        .map_err(|_| FetchError::InvalidResponse)?;
+
+        let data_result = JsFuture::from(data_promise).await;
+        if *self.active.borrow() {
+            data_result
+                .map_err(|_| FetchError::InvalidResponse)
+                .and_then(DATA::from_js)
+        } else {
+            Err(FetchError::Canceled)
+        }
+    }
+}
+
+fn build_request(parts: Parts, body: &JsValue) -> Result<WebRequest, anyhow::Error> {
+    // Map headers into a Js `Header` type.
+    let header_list = parts
+        .headers
+        .iter()
+        .map(|(k, v)| {
+            Ok(Array::from_iter(&[
+                JsValue::from_str(k.as_str()),
+                JsValue::from_str(
+                    v.to_str()
+                        .map_err(|_| anyhow!("Unparsable request header"))?,
+                ),
+            ]))
+        })
+        .collect::<Result<Array, anyhow::Error>>()?;
+
+    let header_map = Headers::new_with_str_sequence_sequence(&header_list)
+        .map_err(|_| anyhow!("couldn't build headers"))?;
```

Comment:
```suggestion
        .context(couldn't build headers")?;
```

Should work here, but needs a `anyhow::Context` import.

Reply:
oh cool!

Reply:
it's late here, so feel free to make changes to my branch or just merge this into yours

Reply:
Will do.